### PR TITLE
[loader] Determine assembly load context from path

### LINKED ISSUE #8575: GetCustomAttributes on a type from an assembly in the LoadFile context fails
## Steps to Reproduce

1. Extract the attached repro
2. `make run`

[Repro.zip](https://github.com/mono/mono/files/1974822/Repro.zip)

### Current Behavior

It throws an Exception.

### Expected Behavior

Console shows `Found 1 MyCustomAttribute attributes`

## On which platforms did you notice this

[ ] macOS
[X ] Linux
[ ] Windows

The repro case gives the expected output on .Net framework and Mono up to 5.12.0.128.
It gives the exception on mono/master.

**Version Used**:

```
$ mono --version
Mono JIT compiler version 5.15.0 (master/dbfbffb559a Fri May  4 13:57:34 CEST 2018)
Copyright (C) 2002-2014 Novell, Inc, Xamarin Inc and Contributors. www.mono-project.com
        TLS:           __thread
        SIGSEGV:       altstack
        Notifications: epoll
        Architecture:  amd64
        Disabled:      none
        Misc:          softdebug
        Interpreter:   yes
        LLVM:          supported, not enabled.
        GC:            sgen (concurrent by default)
$ gcc --version | head -1
gcc (GCC) 4.4.7 20120313 (Red Hat 4.4.7-17)
$ uname -srvm
Linux 2.6.32-696.3.2.el6.x86_64 #1 SMP Wed Jun 7 11:51:39 EDT 2017 x86_64
```

## Stacktrace

```
Can't find custom attr constructor image: /scratch/repro/A.dll mtoken: 0x0a000001 due to: Could not load file or assembly 'Attr, Version=0.0.0.0, Culture=neutral, PublicKeyToken=null' or one of its dependencies.
Got an exception: System.IO.FileNotFoundException: Could not load file or assembly 'Attr, Version=0.0.0.0, Culture=neutral, PublicKeyToken=null' or one of its dependencies.
File name: 'Attr, Version=0.0.0.0, Culture=neutral, PublicKeyToken=null'
  at (wrapper managed-to-native) System.MonoCustomAttrs.GetCustomAttributesDataInternal(System.Reflection.ICustomAttributeProvider)
  at System.MonoCustomAttrs.GetCustomAttributesData (System.Reflection.ICustomAttributeProvider obj) [0x0000e] in /scratch/mono/mono-git/mcs/class/corlib/System/MonoCustomAttrs.cs:303
  at System.Reflection.CustomAttributeData.GetCustomAttributesInternal (System.RuntimeType target) [0x00000] in /scratch/mono/mono-git/mcs/class/corlib/System.Reflection/CustomAttributeData.cs:123
  at System.RuntimeType.GetCustomAttributesData () [0x00000] in /scratch/mono/mono-git/mcs/class/referencesource/mscorlib/system/rttype.cs:5119
  at System.Reflection.MemberInfo.get_CustomAttributes () [0x00000] in /scratch/mono/mono-git/mcs/class/referencesource/mscorlib/system/reflection/memberinfo.cs:51
  at P.Prog.Main () [0x0006e] in <f898189524a448b8a9de610606c5ddab>:0
```

### LINKED ISSUE #9542: Assembly.Load from an msbuild task does not load assemblies from the task's directory
Given a custom msbuild task, `mytask.dll`:

```
using System;
using System.Reflection;
using Microsoft.Build.Utilities;

namespace Test {
	public class MyTask : Task
	{
		public override bool Execute()
		{
			var assemblyName = typeof(MyTask).Assembly.GetName();
			Console.WriteLine ($"** CodeBase for MyTask's assembly name: {assemblyName.CodeBase}");
			assemblyName.Name = "FooDep";
			var assembly = Assembly.Load(assemblyName);	

			return true;
		}
	}
}
```

.. tries to load `FooDep.dll`, located in the same directory using `Assembly.Load`, but that fails on mono with:

```
"/Users/ankit/test/cs/test/test.targets" (default target) (1) ->
(Main target) ->
  /Users/ankit/test/cs/test/test.targets(5,3): error MSB4018: The "MyTask" task failed unexpectedly.
/Users/ankit/test/cs/test/test.targets(5,3): error MSB4018: System.IO.FileNotFoundException: Could not load file or assembly 'FooDep' or one of its dependencies
/Users/ankit/test/cs/test/test.targets(5,3): error MSB4018: File name: 'FooDep'
/Users/ankit/test/cs/test/test.targets(5,3): error MSB4018:   at System.AppDomain.Load (System.Reflection.AssemblyName assemblyRef, System.Security.Policy.Evidence assemblySecurity) [0x001a9] in /Users/builder/jenkins/workspace/build-package-osx-mono/2018-04/external/bockbuild/builds/mono-x64/mcs/class/corlib/System/AppDomain.cs:777
/Users/ankit/test/cs/test/test.targets(5,3): error MSB4018:   at (wrapper remoting-invoke-with-check) System.AppDomain.Load(System.Reflection.AssemblyName,System.Security.Policy.Evidence)
/Users/ankit/test/cs/test/test.targets(5,3): error MSB4018:   at System.AppDomain.Load (System.Reflection.AssemblyName assemblyRef) [0x00000] in /Users/builder/jenkins/workspace/build-package-osx-mono/2018-04/external/bockbuild/builds/mono-x64/mcs/class/corlib/System/AppDomain.cs:714
/Users/ankit/test/cs/test/test.targets(5,3): error MSB4018:   at (wrapper remoting-invoke-with-check) System.AppDomain.Load(System.Reflection.AssemblyName)
/Users/ankit/test/cs/test/test.targets(5,3): error MSB4018:   at System.Reflection.Assembly.Load (System.Reflection.AssemblyName assemblyRef) [0x00000] in /Users/builder/jenkins/workspace/build-package-osx-mono/2018-04/external/bockbuild/builds/mono-x64/mcs/class/corlib/System.Reflection/Assembly.cs:600
/Users/ankit/test/cs/test/test.targets(5,3): error MSB4018:   at Test.MyTask.Execute () [0x00038] in <9f7b837c6f784c6184f711b4d34f6188>:0
/Users/ankit/test/cs/test/test.targets(5,3): error MSB4018:   at Microsoft.Build.BackEnd.TaskExecutionHost.Microsoft.Build.BackEnd.ITaskExecutionHost.Execute () [0x00023] in /_/src/Build/BackEnd/TaskExecutionHost/TaskExecutionHost.cs:631
/Users/ankit/test/cs/test/test.targets(5,3): error MSB4018:   at Microsoft.Build.BackEnd.TaskBuilder+<ExecuteInstantiatedTask>d__26.MoveNext () [0x001f6] in /_/src/Build/BackEnd/Components/RequestBuilder/TaskBuilder.cs:787
```

Running with `MONO_LOG_LEVEL=debug MONO_LOG_MASK=asm` shows:

```
** CodeBase for MyTask's assembly name: file:///Users/ankit/test/cs/test/mytask.dll

Mono: Domain MSBuild.dll search path is:
Mono:   path[0] = '/Library/Frameworks/Mono.framework/Versions/5.14.0/lib/mono/msbuild/15.0/bin/'
Mono: End of domain MSBuild.dll search path.
Mono: Assembly Loader probing location: '/Library/Frameworks/Mono.framework/Versions/5.14.0/lib/FooDep.dll'.
Mono: Assembly Loader probing location: '/Library/Frameworks/Mono.framework/Versions/5.14.0/lib/mono/4.5//Facades/FooDep.dll'.
Mono: Assembly Loader probing location: '/Library/Frameworks/Mono.framework/Versions/5.14.0/lib/FooDep.exe'.
Mono: Assembly Loader probing location: '/Library/Frameworks/Mono.framework/Versions/5.14.0/lib/mono/4.5//Facades/FooDep.exe'.
Mono: Unloading image /Users/ankit/test/cs/test/mytask.dll [0x7fb2c4082600].
Mono: Assembly Loader probing location: '/Users/ankit/test/cs/test/mytask.dll'.
```

Running the same thing on windows, `fuslogvw` shows:

```
LOG: Attempting download of new URL file:///C:/Program Files (x86)/Microsoft Visual Studio/2017/Professional/MSBuild/15.0/bin/FooDep.DLL.
LOG: Attempting download of new URL file:///C:/Program Files (x86)/Microsoft Visual Studio/2017/Professional/MSBuild/15.0/bin/FooDep/FooDep.DLL.
LOG: Attempting download of new URL file:///C:/Program Files (x86)/Microsoft Visual Studio/2017/Professional/MSBuild/15.0/bin/FooDep.EXE.
LOG: Attempting download of new URL file:///C:/Program Files (x86)/Microsoft Visual Studio/2017/Professional/MSBuild/15.0/bin/FooDep/FooDep.EXE.
LOG: Attempting download of new URL file:///C:/Users/ankit/Downloads/task/FooDep.DLL.
LOG: Assembly download was successful. Attempting setup of file: C:\Users\ankit\Downloads\task\FooDep.dll
```

And `** CodeBase for MyTask's assembly name: file:///C:/Users/ankit/Downloads/task/mytask.dll` from msbuild.
And the assembly loading works fine.

This is breaking build for msbuild.

`Mono JIT compiler version 5.14.0.161 (2018-04/c95a4a2d2c5 Tue Jul 10 06:48:32 EDT 2018)`

cc @lambdageek

### LINKED ISSUE #9753: [loader] LoadFile from application base should load in default context
If an assembly is found in the application base directory, even if starting from `LoadFrom` or `LoadFile`, it should go into the default context.  That affects loading of subsequent referenced assemblies:

[LoadAndReference.zip](https://github.com/mono/mono/files/2233363/LoadAndReference.zip)

Here we have `Main.exe` which can do either a `LoadFrom` or a `LoadFile` of `Thing.dll` which, in turn, references `Relative.dll`.

The behavior of .NET Framework observed with FUSLOGVW:

* if all three assemblies are in the same directory, then both Thing.dll and Relative.dll are found and loaded into the default context.
* if Thing.dll and Relative.dll are both in an unrelated directory
  * `LoadFile` will load Thing.dll in an individual load context and subsequently fail to find `Relative.dll` (it's says it's starting from the default context).
  * `LoadFrom` will load Thing.dll in LoadFrom context and subsequently find `Relative.dll` next to Thing.dll and load it also in LoadFrom context.
* if Thing.dll is in an unrelated directory, but Relative.dll is next to Main.dll then:
  * `LoadFrom`  will load Thing.dll in LoadFrom context; then Relative.dll will be searched for starting in LoadFrom context and successfully loaded in the default context since it's in the application base directory.
  * `LoadFile` will load Thing.dll in individual context; then Relative.dll will be searched for starting in the default context and successfully loaded in the default context since it's in the application base directory.

This is kind of related to #9542 in that assemblies in the application base directory, even if loaded using `LoadFrom` or `LoadFile` should load in a default context. (In 9542 we actually _don't_ want to be in the default context, but here we _do_).

### PR DESCRIPTION
**Pull request by lambdageek**
The assembly loader uses assembly load contexts to determine how referenced assemblies and explict alls to `Assembly.Load`, `Assembly.LoadFrom`, and `Assembly.LoadFile` are resolved.

Previously the assembly load context was determined purely from how an assembly load request originated: for referenced assemblies it was determined by the requesting assembly; for `Assembly.Load` it was the default context, for `LoadFrom` the LoadFrom context, and for `LoadFile` the individual (or "no context") assembly context.

However that doesn't match exactly the observed behavior of .NET Framework.

So this PR adds additional complications:

1. If `Assembly.Load` is called by an executing assembly in `LoadFrom` context, try to load the given assembly in LoadFrom context and probe in the executing assembly's base directory.
2. If `Assembly.LoadFrom` or `Assembly.LoadFile` is called with a filename that matches the appdomain search path (the appdomain base directory or the private bin path), load the assembly in the default context.
3. If `Assembly.LoadFrom` or `Assembly.LoadFile` is called with a filename that matches the GAC, load the assembly in the default context.
4. If an assembly was previously loaded into the individual assembly context, we can still load its references if they're in the GAC or MONO_PATH or the appdomain base directory or private bin path.

Fixes #9753  and fixes #9542  and fixes #8575

### CODE REVIEW

**Comment by lambdageek** [comment]
@monojenkins build pkg

**Comment by luhenry** [comment]
How are we testing this?

**Comment by lambdageek** [comment]
working on some tests.  (The main hangup is writing Makefile rules to put all the right pieces in the right subdirectories)

I wanted to build a package to try @alanmcgovern's issue first in case I messed up something on a non-toy workload.

**Comment by alanmcgovern** [comment]
Let me know when there's a package to test :)

**Comment by lambdageek** [comment]
@monojenkins build pkg

**Comment by lambdageek** [comment]
@monojenkins backport to 2018-06

**Comment by marek-safar** [comment]
@kumpera could you review this?

File: mono/metadata/appdomain.c [hunk 1]
```diff
@@ -1412,6 +1416,21 @@ mono_domain_fire_assembly_load (MonoAssembly *assembly, gpointer user_data)
 	HANDLE_FUNCTION_RETURN ();
 }
 
+static gboolean
+mono_domain_asmctx_from_path (const char *fname, MonoAssembly *requesting_assembly, gpointer user_data, MonoAssemblyContextKind *out_asmctx)
+{
+	MonoDomain *domain = mono_domain_get ();
```
Comment (kumpera) [inline mono/metadata/appdomain.c]: It's usually good to always init out vars

**Comment by lambdageek** [comment]
@monojenkins backport to 2018-08

File: mono/metadata/appdomain.c (hunk 1 above)
Comment (lambdageek) [reply mono/metadata/appdomain.c]: The hook invoke function `assembly_invoke_asmctx_from_path_hook` initializes the out var.  The individual hook functions only write to it when they return true.

**Pull request merged**